lcpserver/api/license.go

```diff
@@ -151,43 +152,90 @@ func buildLicense(lic *license.License, s Server) error {
 	return nil
 }
 
+func copyZipFile(out *zip.Writer, in *zip.Reader) error {
+	for _, file := range in.File {
+		newFile, err := out.CreateHeader(&file.FileHeader)
+		if err != nil {
+			return err
+		}
+
+		r, err := file.Open()
+		if err != nil {
+			return err
+		}
+
+		_, err = io.Copy(newFile, r)
+		r.Close()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func isWebPub(in *zip.Reader) bool {
+	for _, f := range in.File {
+		if f.Name == "manifest.json" {
+			return true
+		}
+	}
+
+	return false
+}
+
 // build a licensed publication, common to get and generate licensed publication
 //
-func buildLicencedPublication(lic *license.License, s Server) (ep epub.Epub, err error) {
+func buildLicensedPublication(lic *license.License, s Server) (buf bytes.Buffer, err error) {
 	// get the epub content info from the bd
 	epubFile, err := s.Store().Get(lic.ContentId)
 	if err != nil {
 		return
 	}
 	// get the epub content
-	epubContent, err1 := epubFile.Contents()
+	contents, err1 := epubFile.Contents()
 	if err1 != nil {
-		return ep, err1
+		return buf, err1
 	}
-	var b bytes.Buffer
-	// copy the epub content to a buffer
-	io.Copy(&b, epubContent)
-	// create a zip reader
-	zr, err2 := zip.NewReader(bytes.NewReader(b.Bytes()), int64(b.Len()))
-	if err2 != nil {
-		return ep, err2
+
+	b, err := ioutil.ReadAll(contents)
+	if err != nil {
+		return buf, err
 	}
-	ep, err3 := epub.Read(zr)
-	if err3 != nil {
-		return ep, err3
+	// create a zip reader
+	zr, err := zip.NewReader(bytes.NewReader(b), int64(len(b)))
+	if err != nil {
+		return buf, err
 	}
-	// add the license to publication
-	var buf bytes.Buffer
-	enc := json.NewEncoder(&buf)
-	// do not escape characters
-	enc.SetEscapeHTML(false)
-	enc.Encode(lic)
+
+	zipWriter := zip.NewWriter(&buf)
+	err = copyZipFile(zipWriter, zr)
+
+	// Encode the license to JSON, removing the trailing newline
 	// write the buffer in the zip, and suppress the trailing newline
-	// FIXME: check that the newline is not present anymore
-	// FIXME/ try to optimize with buf.ReadBytes(byte('\n')) instead of creating a new buffer.
-	var buf2 bytes.Buffer
-	buf2.Write(bytes.TrimRight(buf.Bytes(), "\n"))
-	ep.Add(epub.LicenseFile, &buf2, uint64(buf2.Len()))
+	licenseBytes, err := json.Marshal(lic)
+	if err != nil {
+		return buf, err
+	}
+
+	licenseBytes = bytes.TrimRight(licenseBytes, "\n")
+	zipWriter.Close()
```

Comment:
zipWrite is closed here and used later. I will read the doc but it may be the reason why I don't see licenses in packaged protected pdf.

Reply:
Yeah that sounds wrong let me move that around